### DISCUSSION TITLE
Status check for a matrix jobs

### DISCUSSION THREAD

**Discussion by edzis** [community/community#26822] [Actions] [answered]
<p>I build several subprojects using a job with the matrix strategy. The matrix is dynamic - depends on the branch used. All of those matrix jobs need to succeed for the code to be good to merge. But I fail to find a way how to check the composite status of the branches. Which direction would you recommend?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/edzis">@edzis</a>,</p>
<blockquote>
<p>The matrix is dynamic - depends on the branch used.</p>
</blockquote>
<p>Do you mean that you set different matrix configurations in the workflow on different branches?</p>
<blockquote>
<p>But I fail to find a way how to check the composite status of the branches.</p>
</blockquote>
<p>What do you mean about “<strong>the composite status of the branches</strong>”?<br>
Do you mean that summarize the latest checks statuses of all the branches as a composite status?<br>
If so, maybe you can try to use the <strong>repository_dispatch</strong> to trigger a workflow to summarize all the checks statuses.</p>
<p>If I missed something, feel free to tell me. And if possible, you also can share your repository with us, so that we can check more detailed configurations related to the workflow.</p>

**Comment by ylemkimon** [comment]
<p>I’m using a job that <code>needs</code> matrix job to check all matrix jobs are successful. For example:</p>
<pre><code class="lang-yaml">jobs:
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        number: [0, 1, 2]

    steps:
    - run: echo ${{ matrix.number }}
  
  build_success:
    runs-on: ubuntu-latest
    needs: build
    steps:
    - run: echo Done!
</code></pre>
<p>In the example above, <code>build_success</code> job is only run after all <code>build</code> jobs are successful, so you can check on the result of <code>build_success</code> job.</p>

**Comment by edzis** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="127354">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>Do you mean that you set different matrix configurations in the workflow on different branches?</p>
</blockquote>
</aside>
<p>Yes.</p>
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="127354">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>What do you mean about “ <strong>the composite status of the branches</strong> ”?</p>
</blockquote>
</aside>
<p>PR can be merged only if all of the jobs from the calculated matrix complete successfully. If a single job from the matrix fails, the PR merge should be blocked.</p>
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="127354">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>use the  <strong>repository_dispatch</strong>  to trigger a workflow</p>
</blockquote>
</aside>
<p>I don’t understand where and what would I dispatch and how would that help.</p>
<aside class="quote no-group" data-username="ylemkimon" data-post="3" data-topic="127354">
<div class="title">
<div class="quote-controls"></div>
 ylemkimon:</div>
<blockquote>
<p>I’m using a job that <code>needs</code> matrix job to check all matrix jobs are successful.</p>
</blockquote>
</aside>
<p>This gives false-positives - if one of matrix <code>build</code> jobs fail, the <code>build_success</code> is skipped, which is considered as a pass. See screenshot:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/530fc5faa604de47bf4872f3033e3c216387f6ff.png" data-download-href="/uploads/short-url/bQNqj24yJueXqt5f18jWJeLqS2b.png?dl=1" title="Screen Shot 2020-08-18 at 10.50.14" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/163636/181100043-bfb9be14-4f95-4998-9b5c-1dce1c1f3397.png" alt="Screen Shot 2020-08-18 at 10.50.14" data-base62-sha1="bQNqj24yJueXqt5f18jWJeLqS2b" width="690" height="314" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/530fc5faa604de47bf4872f3033e3c216387f6ff_2_690x314.png, https://aws1.discourse-cdn.com/github/optimized/2X/5/530fc5faa604de47bf4872f3033e3c216387f6ff_2_1035x471.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/5/530fc5faa604de47bf4872f3033e3c216387f6ff_2_1380x628.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/530fc5faa604de47bf4872f3033e3c216387f6ff_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-08-18 at 10.50.14</span><span class="informations">1740×792 97.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="127354">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>summarize all the checks statuses</p>
</blockquote>
</aside>
<p>I might generate trivial artifacts from each running matrix job and have a job that checks the existence of those based on matrix, but that sounds over-engineered. Hoping to find something more simple.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/edzis">@edzis</a> ,</p>
<p>The checks of a PR generally include the jobs in the workflows related to the PR, and some checks generated from other GitHub Apps.<br>
The related workflows can be:</p>
<ol>
<li>
<p>triggered by the events related to pull request, such as <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request" rel="nofollow noopener"><strong>pull_request</strong></a>, <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_review" rel="nofollow noopener"><strong>pull_request_review</strong></a>, <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_review_comment" rel="nofollow noopener"><strong>pull_request_review_comment</strong></a> and <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target" rel="nofollow noopener"><strong>pull_request_target</strong></a>.</p>
</li>
<li>
<p>triggered by <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push" rel="nofollow noopener"><strong>push</strong></a> event when pushing new commits to the head branch (source branch) of the PR. The new commits push also triggers the workflows (if any) run on the activity type ‘<strong>synchronize</strong>’ of <strong>pull_request</strong> event if any.</p>
</li>
</ol>
<p>Normally, the workflows run on other branches that not related to the PR will not affect to this PR.</p>
<p>If you want the related checks on the PR must be passed before the PR can be merged, you can set some branch protection rules, and set the related checks as ‘<strong>Required</strong>’. More details, see “<a href="https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests" rel="nofollow noopener"><strong>Defining the mergeability of pull requests</strong></a>”.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/772d1acd87be36f19a121dc89f3f320d39a71a41.png" data-download-href="/uploads/short-url/h0hsenKwz1dCQ1NrmLMerm1Ag0h.png?dl=1" title="required-checks" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181100045-835a31c7-7ea9-4c98-9a80-a17bd1c151dc.png" alt="required-checks" data-base62-sha1="h0hsenKwz1dCQ1NrmLMerm1Ag0h" width="690" height="418" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/772d1acd87be36f19a121dc89f3f320d39a71a41_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">required-checks</span><span class="informations">995×603 37 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by edzis** [comment]
<p><a class="mention" href="/brightran">@brightran</a> I believe we are not on the same page.  I believe it might help for me to share more insights about the setup. Here is a short description and the action setup is shared on <a href="https://gist.github.com/edzis/17602cb75eb6346b9096b7b53498a424" rel="nofollow noopener">https://gist.github.com/edzis/17602cb75eb6346b9096b7b53498a424</a></p>
<p>Conceptually we have a monorepo with multiple projects (customizers) and a shared codebase used by those projects.<br>
We calculate a matrix based on branch name.<br>
When we work on a specific project, we do the work in a branch named <code>CLIENT-PROJECT/FEATURE</code>. Only that project needs to be tested and deployed on dev - the matrix contains only that project and the dev environment.<br>
When we work on generic features only the tests need to be run but there should be no deployment  - the matrix job defines <code>deploy = false</code> and only the <code>onlyTest</code> job is run.<br>
When a PR is merged into master we need to test and deploy all projects to dev - matrix includes all projects and  the dev environment.<br>
When a project needs to be deployed to staging or production, we push a <code>staging/CLIENT-PROJECT</code> or <code>production/CLIENT-PROJECT</code> branch - the matrix contains only that project and the specified environment.</p>
<p>Any ideas how I can check if all the jobs defined by the matrix have succeeded?</p>
<ul>
<li>I can not specify the parent “Deploy” job as required because that is newer run from the perspective of merge guards.</li>
<li>I can not specify all the possible jobs resulting from the matrix as most of them are not even run in feature branches.</li>
<li>I can not specify another job that uses the build job as required because, if one of my matrix jobs fails, that job is a false positive - it is skipped which is considered as a pass by the merge guard.</li>
</ul>

**Comment by hodbn** [comment]
<p>Hi <a class="mention" href="/edzis">@edzis</a>,<br>
IIUIC you want to summarize a matrix job to a single check. This way, you can tick only <em>one</em> checkbox in the protected branch settings of the repository and later matrix entries will only change in the corresponding workflow file.</p>
<p>This is a workaround using a utility job (<code>buildall</code>) after your matrix job (<code>build</code>):</p>
<pre><code class="lang-yaml">  buildall:
    if: ${{ always() }}
    runs-on: ubuntu-latest
    name: Build (matrix)
    needs: build
    steps:
      - name: Check build matrix status
        if: ${{ needs.build.result != 'success' }}
        run: exit 1
</code></pre>
<p>In addition, I think you should have <code>continue-on-error: false</code> on the <code>build</code> job (it’s the default).</p>

**Comment by edzis** [comment]
<p>Thanks <a class="mention" href="/hodbn">@hodbn</a>, that fixes my issue!</p>
<p>I was not aware of</p>
<pre><code>if: ${{ always() }}
</code></pre>
<p>and</p>
<pre><code>if: ${{ needs.build.result != 'success' }}
</code></pre>
<p>And I tried <code>continue-on-error: false</code> before with no success, but kept the current setup because when merging into master I don’t want to stop all projects when some fail.</p>

**Comment by dudicoco** [comment]
<p><a class="mention" href="/hodbn">@hodbn</a> <a class="mention" href="/edzis">@edzis</a> this solution does not seem to work.<br>
If one of the matrix jobs failed while another succeeded, <code>needs.build.result</code> value will be <code>success</code>.</p>

**Reply by thaysealencar** [reply]
Same for me. I keep getting `success` even though one of the jobs of my matrix failed.

**Comment by egmacke** [comment]
<p>Is there any update on this?</p>
<p><a class="mention" href="/dudicoco">@dudicoco</a> is correct that if any of the matrix jobs are successful then <code>needs.build.result</code> will be <code>success</code></p>
<p>I have a matrix build that builds any projects that have changed in our mono-repo. I then want to have a single job that runs after so I can configure the github branch protection to require that single job, which would accumulate the results of the matrix job (i.e. if any of the matrix jobs fail, it would also fail)</p>
<p>I could achieve this by turning off <code>fail-fast</code> but that would be extremely couter productive as projects passing/failing build&amp;test are typically independent of eachother.</p>

**Comment by dudicoco** [comment]
<p><a class="mention" href="/egmacke">@egmacke</a> I was able to solve this using the following suggestion: <a href="https://github.community/t/get-status-of-parallel-jobs-into-a-single-webhook-payload/18077/2" class="inline-onebox">Get status of parallel jobs into a single Webhook payload - #2 by weide-zhou</a></p>

**Reply by thaysealencar** [reply]
@dudicoco Thanks so much for sharing the link to this discussion. The solution proposed there with the GH API worked perfectly for me.

**Comment by egmacke** [comment]
<p><a class="mention" href="/dudicoco">@dudicoco</a> thanks! That looks like a nice solution.</p>
<p>However, since posting, I’ve come up with a different alternative, which is to add an additional step in the matrix job that adds a tag to the pull request if it fails. Then I have a final job that checks if the tag has been added or not.</p>
<p>I’ll have a look and see which approach is better for my use case, but thanks for the link!</p>
<p>For reference for anyone else coming here:</p>
<pre><code class="lang-auto">name: CI Checks

on:
  pull_request:
    branches: [master]
    types: [opened, synchronize, reopened]
  push:
    branches: [master]

jobs:
  changes:
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.filter.outputs.changes }}
    steps:
      - uses: actions/checkout@v2

      - uses: dorny/paths-filter@v2
        id: filter
        with:
          filters: '.github/project-filters.yml'

      - name: Clean up label
        uses: actions-ecosystem/action-remove-labels@v1
        if: success() &amp;&amp; github.event.pull_request
        with:
          labels: build-fail
          fail_on_error: false

  ci-checks:
    name: CI Checks
    runs-on: ubuntu-latest
    needs: changes
    if: fromJSON(needs.changes.outputs.matrix)[0] != null
    strategy:
      fail-fast: false
      matrix: 
        project: ${{ fromJSON(needs.changes.outputs.matrix) }}
    steps:
      - uses: actions/checkout@v2

      - name: Do build/test stuff here
        run: echo "Do build"

      - name: Update label
        uses: actions-ecosystem/action-add-labels@v1
        if: failure()
        with:
          labels: build-fail

  ci-build-check:
    if: always() &amp;&amp; github.event.pull_request
    runs-on: ubuntu-latest
    name: CI Build Check
    needs: 
      - ci-checks
      - changes
    steps:
      - uses: mheap/github-action-required-labels@v1
        with:
          mode: exactly
          count: 0
          labels: build-fail
</code></pre>

**Comment by richja** [comment, marked as answer]
late to the party but looks like `result` of matrix now works as expected when using: `needs.<job_id>.result`

so I can have just one single additional job which I can set as required
````
results:
    if: ${{ always() }}
    runs-on: ubuntu-latest
    name: Final Results
    needs: [build]
    steps:
      - run: |
          result="${{ needs.build.result }}"
          if [[ $result == "success" || $result == "skipped" ]]; then
            exit 0
          else
            exit 1
          fi
````

**Reply by nkuehn** [reply]
Be it for a change on GitHub or not, this is the best recommendation in practice because as opposed to the pinned answer, the "fake" step is always present with a result and not only if failed.

**Reply by radoslaw-szpiech** [reply]
This is very cool solution with one "but". If no jobs are run in matrix (in my case it's a list of affected apps which can be empty), this job will be marked as failed. Any idea why?

EDIT: Nvm, just needed to add `if` to matrix job. ;)

**Reply by sounisi5011** [reply]
This condition can be made simpler by using [the `contains()` function](https://docs.github.com/en/actions/learn-github-actions/expressions#contains).

```yaml
jobs:
  build:
    ...

  results:
    if: ${{ always() }}
    runs-on: ubuntu-latest
    name: Final Results
    needs: [build]
    steps:
      - run: exit 1
        # see https://stackoverflow.com/a/67532120/4907315
        if: >-
          ${{
               contains(needs.*.result, 'failure')
            || contains(needs.*.result, 'cancelled')
          }}
```

This job uses `needs.*.result` expressions, so it is easy to update it to check other jobs later. Just add to `jobs.<job_id>.needs`!

<details><summary>Show example of diff</summary>

```diff
   results:
     if: ${{ always() }}
     runs-on: ubuntu-latest
     name: Final Results
-    needs: [build]
+    needs: [build, lint, test]
     steps:
       - run: exit 1
         # see https://stackoverflow.com/a/67532120/4907315
         if: >-
           ${{
                contains(needs.*.result, 'failure')
             || contains(needs.*.result, 'cancelled')
           }}
```

</details>

I using this job [in my repository](https://github.com/sounisi5011/npm-packages/blob/2a5ca2de696eeb8b40a38de90580441c4c6c96e0/.github/workflows/ci.yaml#L482-L498). So far, it is working as expected.

**Reply by etianen** [reply]
That's really neat, @sounisi5011!
I think it can be improved further by failing on the `'skipped'` result:

``` yaml
jobs:
  build:
    ...

  results:
    if: ${{ always() }}
    runs-on: ubuntu-latest
    name: Final Results
    needs: [build]
    steps:
      - run: exit 1
        # see https://stackoverflow.com/a/67532120/4907315
        if: >-
          ${{
               contains(needs.*.result, 'failure')
            || contains(needs.*.result, 'cancelled')
            || contains(needs.*.result, 'skipped')
          }}
```

Example [workflow file](https://github.com/etianen/logot/blob/main/.github/workflows/build.yml).

**Reply by krishna13052001** [reply]
@sounisi5011 , is there a way to print no of jobs failed too?

**Reply by sounisi5011** [reply]
> @sounisi5011 , is there a way to print no of jobs failed too?

@krishna13052001, we probably can't count the elements in an array (in this case, `needs.*.result`) using only [the built-in functions in the GitHub Workflow](https://docs.github.com/en/actions/learn-github-actions/expressions#functions). So it might be better to convert the array to a string value with the [`join( array, optionalSeparator )`](https://docs.github.com/en/actions/learn-github-actions/expressions#join) or [`toJSON( value )`](https://docs.github.com/en/actions/learn-github-actions/expressions#join) functions, then parse it in Bash or Python and count it[^1]. Filtering of failed jobs is also better done within those scripts.

[^1]: The script specified for [the `run` keyword](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsrun) can be Bash, PowerShell, Python, etc. See [the `shell` parameter](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_iddefaultsrunshell).

**Reply by StephenHodgson** [reply]
I'm not sure I agree this should be the answer. 

I think the actual runner itself should give a more generalized status 'success' result, and we shouldn't have to add an additional step just to work around this. Should be fairly trivial to add the logic in the Runner implementation itself. It'll likely be similar to how [it decides if a "composite" action is successful](https://github.com/actions/runner/blob/7e84ae0b30ceb0c0a00a9194b5cdca2c004711c0/src/Runner.Worker/Expressions/SuccessFunction.cs#L28). 

```csharp
namespace GitHub.Runner.Worker.Expressions
{
    public sealed class SuccessFunction : Function
    {
        protected sealed override object EvaluateCore(EvaluationContext evaluationContext, out ResultMemory resultMemory)
        {
            resultMemory = null;
            var templateContext = evaluationContext.State as TemplateContext;
            ArgUtil.NotNull(templateContext, nameof(templateContext));
            var executionContext = templateContext.State[nameof(IExecutionContext)] as IExecutionContext;
            ArgUtil.NotNull(executionContext, nameof(executionContext));

            // Decide based on 'action_status' for composite MAIN steps and 'job.status' for pre, post and job-level steps
            var isCompositeMainStep = executionContext.IsEmbedded && executionContext.Stage == ActionRunStage.Main;
            if (isCompositeMainStep)
            {
                ActionResult actionStatus = EnumUtil.TryParse<ActionResult>(executionContext.GetGitHubContext("action_status")) ?? ActionResult.Success;
                return actionStatus == ActionResult.Success;
            }
            else
            {
                ActionResult jobStatus = executionContext.JobContext.Status ?? ActionResult.Success;
                return jobStatus == ActionResult.Success;
            }
        }
    }
}
```

**Reply by StephenHodgson** [reply]
Digging into this a bit more. [It seems that the results are only merged into the larger job context if any single step fails](https://github.com/actions/runner/blob/7e84ae0b30ceb0c0a00a9194b5cdca2c004711c0/src/Runner.Worker/StepsRunner.cs#L245-L256):

```csharp
                // Update the job result
                if (step.ExecutionContext.Result == TaskResult.Failed)
                {
                    Trace.Info($"Update job result with current step result '{step.ExecutionContext.Result}'.");
                    jobContext.Result = TaskResultUtil.MergeTaskResults(jobContext.Result, step.ExecutionContext.Result.Value);
                    jobContext.JobContext.Status = jobContext.Result?.ToActionResult();
                }
                else
                {
                    Trace.Info($"No need for updating job result with current step result '{step.ExecutionContext.Result}'.");
                }

                Trace.Info($"Current state: job state = '{jobContext.Result}'");
```

**Reply by StephenHodgson** [reply]
It could also be that they're only listening to ActionsStepTelemetry being sent and never sends the same kind of telemetry for successful Job completion. In either case it'd be nice to have that as well.

**Reply by ju-manns** [reply]
This does not seem to work either. If multiple jobs of the matrix fail and a user reruns a single job in that matrix and that single job succeeds, the results job will also be successful.

**Comment by ju-manns** [comment]
@richja's solution has the flaw that manually rerunning a single matrix job (in the case of multiple failures) will make the results job pass as long as that single job passes.
I see two ways to fix this:

1. Add an aggregation job before the results job. This job only runs if the matrix job ran. Then check if the matrix job was skipped or if the aggregate job was successful.
```
aggregate:
    runs-on: ubuntu-latest
    needs: build
    steps:
      - run: exit 0

results:
    runs-on: ubuntu-latest
    needs: [build, aggregate]
    if: ${{ !cancelled() }}
    steps:
      - if: ${{ needs.build.result != 'skipped' && needs.aggregate.result != 'success' }}
        run: exit 1
      - run: exit 0
```

2. Remove the `if: ${{ always() }}` condition from the results job:
```
results:
    runs-on: ubuntu-latest
    name: Final Results
    needs: [build]
    steps:
      - run: exit 0
```
The second option has the flaw that the status check is only fufilled if the matrix job actually runs. A dynamically created matrix job might have an empty matrix though and adding a dummy job that always runs to each matrix job is cumbersome.

All of this is pretty hacky. Github actions needs an internal solution for this.
